src/util/small_shared_ptr.h

```diff
@@ -0,0 +1,228 @@
+/*******************************************************************\
+
+Module: Small intrusive shared pointers
+
+Author: Reuben Thomas, reuben.thomas@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_SMALL_SHARED_PTR_H
+#define CPROVER_UTIL_SMALL_SHARED_PTR_H
+
+#include <iosfwd>  // ostream
+#include <utility> // swap
+
+// TODO We should liberally scatter `constexpr`s and `noexcept`s on platforms
+// that support them.
+
+// This class is really similar to boost's intrusive_pointer, but can be a bit
+// simpler seeing as we're only using it one place...
+
+template <typename T>
+class small_shared_ptrt final
+{
+public:
+  small_shared_ptrt() = default;
+
+  explicit small_shared_ptrt(T *t) : t_(t)
+  {
+    if(t_)
+    {
+      pointee_increment_use_count(*t_);
+    }
+  }
+
+  small_shared_ptrt(const small_shared_ptrt &rhs) : t_(rhs.t_)
+  {
+    if(t_)
+    {
+      pointee_increment_use_count(*t_);
+    }
+  }
+
+  small_shared_ptrt &operator=(const small_shared_ptrt &rhs)
+  {
+    auto copy(rhs);
+    swap(copy);
+    return *this;
+  }
+
+  small_shared_ptrt(small_shared_ptrt &&rhs)
+  {
+    swap(rhs);
+  }
+
+  small_shared_ptrt &operator=(small_shared_ptrt &&rhs)
+  {
+    swap(rhs);
+    return *this;
+  }
+
+  ~small_shared_ptrt()
+  {
+    if(!t_)
+    {
+      return;
+    }
+    if(pointee_use_count(*t_) == 1)
+    {
+      delete t_;
+      return;
+    }
+    pointee_decrement_use_count(*t_);
+  }
+
+  void swap(small_shared_ptrt &rhs)
+  {
+    std::swap(t_, rhs.t_);
+  }
+
+  T *get() const
+  {
+    return t_;
+  }
+
+  T &operator*() const
+  {
+    return *t_;
+  }
+
+  T *operator->() const
+  {
+    return t_;
+  }
+
+  auto use_count() const -> decltype(pointee_use_count(std::declval<T>()))
+  {
+    return t_ ? pointee_use_count(*t_) : 0;
+  }
+
+  explicit operator bool() const
+  {
+    return t_ != nullptr;
+  }
+
+private:
+  T *t_ = nullptr;
+};
+
+template <typename T>
+std::ostream &operator<<(std::ostream &os, const small_shared_ptrt<T> &ptr)
+{
+  return os << ptr.get();
+}
+
+template <typename T, typename... Ts>
+small_shared_ptrt<T> make_small_shared_ptr(Ts &&... ts)
+{
+  return small_shared_ptrt<T>(new T(std::forward<Ts>(ts)...));
+}
```

Comment:
I feel I've asked this question before, but I don't recall the answer -- maybe you can provide the answer in form of a code comment: shouldn't the constructor of `small_shared_ptrt be hidden to enforce the use of this function?

Reply:
We could do, but that would be inconsistent with unique_ptr and shared_ptr. Additionally, that would mean that make_small_shared_ptr would have to be a static member function of small_shared_ptr (verbose to use) or a friend (very tight coupling).